src/core/umount.c

```diff
@@ -381,6 +383,112 @@ static bool nonunmountable_path(const char *path) {
                 || path_startswith(path, "/run/initramfs");
 }
 
+static int wait_for_child(pid_t pid) {
+        sigset_t mask;
+        int r, status;
+        usec_t until;
+
+        assert_se(sigemptyset(&mask) == 0);
+        assert_se(sigaddset(&mask, SIGCHLD) == 0);
+
+        /* Drop into a sigtimewait-based timeout. Waiting for the
+         * remount pid to exit. */
+        until = now(CLOCK_MONOTONIC) + DEFAULT_TIMEOUT_USEC;
+        for (;;) {
+                usec_t n;
+                struct timespec ts;
+
+                n = now(CLOCK_MONOTONIC);
+                if (n >= until)
+                        break;
+
+                r = sigtimedwait(&mask, NULL, timespec_store(&ts, until - n));
+                if (r < 0) {
+                        if (errno == EAGAIN) {
+                                /* Timed out, remount child is likely hung,
+                                 * but we can try. */
+                                if (waitpid(pid, &status, WNOHANG) == pid)
```

Comment:
not that it matters much, but i'd always use waitid() rather than waitpid() for this, since it's more precise in what events you want to wait for, i.e. just WEXITED...